Local ARS HLD

VladimirKuk opened the pull request:
This document provides high level design for the feature local Adaptive Routing and Switching (ARS).

PRs:
| Repository | PR Title | State | Context |
| ---------- | -------- | ----- | ------- |
| sonic-buildimage | [`Local ARS (Adaptive Routing and Switching)`](https://github.com/sonic-net/sonic-buildimage/pull/22264) | [![PR Status](https://img.shields.io/github/pulls/detail/state/sonic-net/sonic-buildimage/22264)](https://img.shields.io/github/pulls/detail/state/sonic-net/sonic-buildimage/22264) | [![PR Checks](https://img.shields.io/github/status/contexts/pulls/sonic-net/sonic-buildimage/22264)](https://img.shields.io/github/status/contexts/pulls/sonic-net/sonic-buildimage/22264) |
| sonic-swss | [`Local ARS (Adaptive Routing and Switching)`](https://github.com/sonic-net/sonic-swss/pull/3597) | [![PR Status](https://img.shields.io/github/pulls/detail/state/sonic-net/sonic-swss/3597)](https://img.shields.io/github/pulls/detail/state/sonic-net/sonic-swss/3597) | [![PR Checks](https://img.shields.io/github/status/contexts/pulls/sonic-net/sonic-swss/3597)](https://img.shields.io/github/status/contexts/pulls/sonic-net/sonic-swss/3597) |
| sonic-swss-common | [`Local ARS (Adaptive Routing and Switching)`](https://github.com/sonic-net/sonic-swss-common/pull/1002) | [![PR Status](https://img.shields.io/github/pulls/detail/state/sonic-net/sonic-swss-common/1002)](https://img.shields.io/github/pulls/detail/state/sonic-net/sonic-swss-common/1002) | [![PR Checks](https://img.shields.io/github/status/contexts/pulls/sonic-net/sonic-swss-common/1002)](https://img.shields.io/github/status/contexts/pulls/sonic-net/sonic-swss-common/1002) |
| sonic-mgmt | [`Local ARS (Adaptive Routing and Switching test plan)`](https://github.com/sonic-net/sonic-mgmt/pull/21237) | [![PR Status](https://img.shields.io/github/pulls/detail/state/sonic-net/sonic-mgmt/21237)](https://img.shields.io/github/pulls/detail/state/sonic-net/sonic-mgmt/21237) | [![PR Checks](https://img.shields.io/github/status/contexts/pulls/sonic-net/sonic-mgmt/21237)](https://img.shields.io/github/status/contexts/pulls/sonic-net/sonic-mgmt/21237) |
| sonic-mgmt | [`Local ARS (Adaptive Routing and Switching test plan)`](https://github.com/sonic-net/sonic-mgmt/pull/24845) | [![PR Status](https://img.shields.io/github/pulls/detail/state/sonic-net/sonic-mgmt/24845)](https://img.shields.io/github/pulls/detail/state/sonic-net/sonic-mgmt/24845) | [![PR Checks](https://img.shields.io/github/status/contexts/pulls/sonic-net/sonic-mgmt/24845)](https://img.shields.io/github/status/contexts/pulls/sonic-net/sonic-mgmt/24845) |

VladimirKuk committed:
initial HLD version

VladimirKuk committed:
review changes

Signed-off-by: Vladimir Kuk <vkuk@marvell.com>

VladimirKuk committed:
Identation fix

Signed-off-by: Vladimir Kuk <vkuk@marvell.com>

VladimirKuk committed:
Identation fix

Signed-off-by: Vladimir Kuk <vkuk@marvell.com>

VladimirKuk committed:
changes after initial review

VladimirKuk committed:
update fonts

VladimirKuk committed:
intergrated ARS_OBJECT in NHG and LAG, use notification mechanism

VladimirKuk committed:
removed ARS_OBJECT

VladimirKuk committed:
Update to latest changes

Signed-off-by: Vladimir Kuk <vkuk@marvell.com>

VladimirKuk committed:
Review changes

Signed-off-by: Vladimir Kuk <vkuk@marvell.com>

VladimirKuk committed:
Addressing comments

Fixed phases contents
Fixed spelling
Added user configured port scaling factor
Updated init sequence
Limit idle time value

Signed-off-by: Vladimir Kuk <vkuk@marvell.com>

VladimirKuk committed:
Addressing comments

Fixed phases contents
Fixed spelling
Added user configured port scaling factor
Updated init sequence
Limit idle time value

Signed-off-by: Vladimir Kuk <vkuk@marvell.com>

mssonicbld commented:
/azp run

Yubin-Li commented:
Hello, why ARS NHG key is prefix? how to handle if there have so many route needs point to ARS ECMP
I think we need to set per port ARS enable bit in Orch
So that if ECMP created by route orch and it's alias belong to enable list, 
We bind ars to this ECMP

Yubin-Li commented:
Why this schema used prefix for Key? dose it means that only some prefix can used with AR ECMP?

VladimirKuk commented:
> Why this schema used prefix for Key? dose it means that only some prefix can used with AR ECMP?

Yes, that is to support use-case for L3 traffic sent to NHG.
Also, currently, there is no other way to reference/identify NHG via CONFIG_DB.

Yubin-Li commented:
> > Why this schema used prefix for Key? dose it means that only some prefix can used with AR ECMP?
> 
> Yes, that is to support use-case for L3 traffic sent to NHG. Also, currently, there is no other way to reference/identify NHG via CONFIG_DB.

If so, how to handle if there have 100 prefixes but point to same ECMP, we want to used AR for this ECMP. It's very hard to set dynamic route point to AR ecmp

I think may be we can define some AR ECMP include port list and trip threshold
for example, I define [port1,2,3,4,5] with AR ecmp, and trip threshold = 4
once we create ECMP with [1,2,3,4] [2,3,4,5], [1,2,3,4,5], we can create with AR

VladimirKuk commented:
> If so, how to handle if there have 100 prefixes but point to same ECMP, we want to used AR for this ECMP. It's very hard to set dynamic route point to AR ecmp
> 
> I think may be we can define some AR ECMP include port list and trip threshold for example, I define [port1,2,3,4,5] with AR ecmp, and trip threshold = 4 once we create ECMP with [1,2,3,4] [2,3,4,5], [1,2,3,4,5], we can create with AR


There is already ARS-enabled interface table, but it could include all interfaces, so I guess shouldn't be used for this. However, to create another table with ARS-enabled interfaces, feels very wasteful.

The problematic scenario is this : suppose you have two NHGs : NHG1 (NH1,NH2,NH3) and NHG2 (NH1,NH2,NH4).
NH3 and NH4 are not reliable and going up and down.
If you rely on ECMP updates to determine whether this is ARS NHG, you could mistake NHG1 with NHG2, since their content will be the same  (NH1,NH2).
Prefix (while it is not ideal) is also, specifies the path you want to enable ARS for.

Also, the use-case was to enable static configuration, To enable dynamic NHG identification, the ARS/NHG ids should, probably, be derived from/imported by some protocol (i.e. BGP).

eddyk-nvidia commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,892 @@
+    - Flowlet-based port selection
+    - Per packet port selection
+2. Support enabling ARS over NHG
+3. Support path quality configuration
```
Please make sure that the design supports 2 mode of operation regarding creation of NHG: done by Route OA and done by FRR when it decides on creation and push the NHG to APP DB

eddyk-nvidia commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,892 @@
+3. Nexthop group's SAI attributes handling:
+    * If nexthop group's ARS object ID, "set" capability is supported:
+        - The nexthop group is updated with the ARS object ID.
+    * If "set" capability is not supported - implement via "create" functionality:
```
what is exactly the trigger for creation of ARS ECMP for both cases - NHG is created either by Route OA and by FRR/Zebra ? Match of the prefix for a created NHG with configured in ARS_NEXTHOP_GROUP table ? On fabric init along the convergence when links go up the Route OA creates multiple NHG (deleting old one and creating a new one) incrementally adding new NHs. So, each NHG will have its own SAI OID. Each one of NHG object will need to be recreated with ARS OID ?

What about check if a port with RIF on that used as NH is ARS-enabled ? You can create now the NHG as ARS-capable with NHs (using RIFs over non-ARS enabled ports). You can think about logic for decision of enabling ARS on NHG considering whether all ports (RIF/NHs) involved in NHG are ARS-enabled or not. In this case non-ARS enabled port (RIF/NH) can't be added to ARS NHG

eddyk-nvidia commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,892 @@
+
+__Figure 9:ARS LAG table creation flow__
+![](images/ars_lag_create.png)
+
```
Please make sure that ARS on LAG is wrapped with the SAI capability check. Some vendors can support ARS for NHG but not for LAG

eddyk-nvidia commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,892 @@
+                type enumeration {
+                    enum ewma {
+                        description "Exponentially Weighted Moving Average algorithm";
+                    }
```
Can potentially expanded with algos supported by other vendors

eddyk-nvidia commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,892 @@
+;field                  = value
+
+algorithm               = "ewma"        ;Path quality calculation algorithm
+max_flows               = uint32        ;Maximum number of flows that can be maintained for ARS
```
Not the only algo used by vendors. Probably will require other vendors to introduce changes in SAI API

eddyk-nvidia commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,892 @@
+    "Ethernet0" : {
+        "scaling_factor": "100"
+    },
+    "Ethernet8" : {}
```
without parameter will be used with default value for "scaling factor" or without it (if not used by other vendor's algo) ?

eddyk-nvidia commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,892 @@
+1. Add ARS_NEXTHOP_GROUP Table
+    * The CONFIG_DB adds the ARS_NEXTHOP_GROUP table, triggering the process.
+2. Create SAI ARS Object
+    * The ArsOrch component receives the update and creates the SAI ARS object.
```
does design propose the creation of ARS Object per entry in ARS_NEXTHOP_GROUP table ? How to re-use the ARS Object ? SAI APIs doesn't limit the re-use

eddyk-nvidia commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,892 @@
+        "max_flows" : "512",
+        "primary_path_threshold" : "100",
+        "alternative_path_cost": "250",
+        "alternative_path_members": {"1.1.1.1", "2.2.2.2"}
```
how do you in advance which NHs can be used when NHG is created from BGP by Route OA or FRR/Zebra ?

VladimirKuk replied:
This should known to user

eddyk-nvidia commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,892 @@
+
+2. Enabling ARS on Ports
+    * Ensure ARS is enabled on specified ports.
+    * Validate behavior when enabling ARS on unsupported interfaces.
```
via enabling ARS on NHG when it is defined on not-supported interfaces ? Missing logic in the design

rajendrat commented:
> > Why this schema used prefix for Key? dose it means that only some prefix can used with AR ECMP?
> 
> Yes, that is to support use-case for L3 traffic sent to NHG. Also, currently, there is no other way to reference/identify NHG via CONFIG_DB.

@VladimirKuk : Is there an option/way to enable ARS for all NH Groups  or enable only for matching the "vrf/ip-prefix"?

ashutosh-agrawal commented on doc/ARS/Local_ARS_HLD.md:
````diff
@@ -0,0 +1,892 @@
+##### ARS_NEXTHOP_GROUP
+
+```
+    container ARS_NEXTHOP_GROUP {
````
ARS_PREFIX could be a better name since key is an IP Prefix and multiple prefixes can share the same nexthop group.

ashutosh-agrawal commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,892 @@
+                description  "Idle duration in microseconds. This duration is to classifying a flow-let in a macro flow.";
+            }
+
+            leaf max_flows {
```
max_flows is a field in the ars_profile table as well. Is this a duplicate or the intent here is to create a separate limit per ip-prefix?

ashutosh-agrawal commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,892 @@
+                description "ARS-enabled interface name";
+            }
+
+            leaf scaling_factor {
```
In figure 5 above, scaling_factor is set by the PortOrch when the port comes up. Is this configuration expected to override the scaling_factor set by PortOrch?

ashutosh-agrawal commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,892 @@
+The ArsOrch component binds to relevant events to start monitoring and processing ARS-related interfaces.
+
+2. Get ARS SAI Capabilities
+ArsOrch retrieves ARS capabilities from the SAI layer and saves them in STATE_DB.
```
Please add a schema for STATE_DB to list what all capabilities are being saved in the DB.

VladimirKuk committed:
Addressing community comments

Separated NHG identification and ARS object
Added NHG matching by ARS-enabled members
Support NH creation from RouteOrch and NhgOrch

Signed-off-by: Vladimir Kuk <vkuk@marvell.com>

mssonicbld commented:
/azp run

VladimirKuk commented:
> > > Why this schema used prefix for Key? dose it means that only some prefix can used with AR ECMP?
> > 
> > 
> > Yes, that is to support use-case for L3 traffic sent to NHG. Also, currently, there is no other way to reference/identify NHG via CONFIG_DB.
> 
> @VladimirKuk : Is there an option/way to enable ARS for all NH Groups or enable only for matching the "vrf/ip-prefix"?

Change matching using nexthop set instead of prefix-based. 
If match-all is introduced, then all the NHGs will be using the same ARS OID, which is not the intended behavior.

mssonicbld commented:
/azp run

VladimirKuk committed:
Fixing Review comments

Signed-off-by: Ashok Kumar P <apannerselva@marvell.com>

mssonicbld commented:
/azp run

venkatmahalingam commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,1052 @@
+
+| SAI api | Supported SAI attribute 
+| ------- | ----------------------- 
+|create_ars_profile | SAI_ARS_PROFILE_ATTR_PORT_LOAD_PAST<br>SAI_ARS_PROFILE_ATTR_LOAD_PAST_MIN_VAL<br>SAI_ARS_PROFILE_ATTR_LOAD_PAST_MAX_VAL<br>SAI_ARS_PROFILE_ATTR_ENABLE_IPV4<br>SAI_ARS_PROFILE_ATTR_ENABLE_IPV6<br>SAI_ARS_PROFILE_ATTR_MAX_FLOWS<br>SAI_ARS_PROFILE_ATTR_ALGO<br>SAI_ARS_PROFILE_ATTR_PORT_LOAD_FUTURE<br>SAI_ARS_PROFILE_ATTR_PORT_LOAD_PAST_WEIGHT<br>SAI_ARS_PROFILE_ATTR_PORT_LOAD_F ...
```
How's the ARS profile binding to a switch handled? refer the section 6.2 in ARS SAI spec.
https://github.com/opencomputeproject/SAI/blob/master/doc/ARS/Adaptive-Routing-and-Switching.md#62-bind-ars-profile-object-to-switch-object

venkatmahalingam commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,1052 @@
+                description "ARS profile Name";
+                mandatory true;
+                type leafref {
+                    path "/sars:sonic-ars/sars:ARS_PROFILE/sars:ARS_PROFILE_LIST/sars:profile_name";
```
Why do we need this reference? is this mentioned in ARS SAI spec?

Yubin-Li commented:
> > If so, how to handle if there have 100 prefixes but point to same ECMP, we want to used AR for this ECMP. It's very hard to set dynamic route point to AR ecmp
> > I think may be we can define some AR ECMP include port list and trip threshold for example, I define [port1,2,3,4,5] with AR ecmp, and trip threshold = 4 once we create ECMP with [1,2,3,4] [2,3,4,5], [1,2,3,4,5], we can create with AR
> 
> There is already ARS-enabled interface table, but it could include all interfaces, so I guess shouldn't be used for this. However, to create another table with ARS-enabled interfaces, feels very wasteful.
> 
> The problematic scenario is this : suppose you have two NHGs : NHG1 (NH1,NH2,NH3) and NHG2 (NH1,NH2,NH4). NH3 and NH4 are not reliable and going up and down. If you rely on ECMP updates to determine whether this is ARS NHG, you could mistake NHG1 with NHG2, since their content will be the same (NH1,NH2). Prefix (while it is not ideal) is also, specifies the path you want to enable ARS for.
> 
> Also, the use-case was to enable static configuration, To enable dynamic NHG identification, the ARS/NHG ids should, probably, be derived from/imported by some protocol (i.e. BGP).

I think for our customer it's very hard to control by using prefixes, 
And I think bgp will not doing this.

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

venkatmahalingam commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,1155 @@
+                default "interface";
+            }
+
+            leaf ars_lag_path_selector_mode {
```
Can we keep only supported fields in the YANG for phase-1?

venkatmahalingam commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,1155 @@
+1. Add ROUTE_TABLE Table
+    * The APPL_DB adds the ROUTE_TABLE table, triggering the process.
+2. Notification
+    * RouteOrch component will notify ArsOrch of NHG change.
```
We talked about removing the notifications between orchagents rather call APIs for any handling? is this not taken care yet?

venkatmahalingam replied:
OK

mssonicbld commented:
/azp run

baxia-lan commented:
Is this HLD ready to merge?

baxia-lan commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,1151 @@
+The ars_nhg_path_selector_mode field in ARS_PROFILE defines the scope at which ARS selection for NH groups is applied.
+The system supports three distinct modes:
+
+**1. Global**:
```
Is there such knob in SAI to enable ARS globally and avoid individual ARS interfaces/objects creation?

venkatmahalingam commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,1151 @@
+
+1. LAG creation
+    * PortsOrch receives LAG from APPL_DB, triggering the process.
+2. Notification
```
Do we still have event notification between orchagent modules? as discussed in the meeting earlier, we can simply make function calls among orchagent modules. please fix it.

sudheer-nexthop commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,1151 @@
+
+3. ARS_INTERFACES holds list of L2 egress interfaces which are ARS-enabled. NHG will be considered ARS-enabled when all of its associated nexthop interfaces are present in the configured ARS_INTERFACES when the match mode configured in ARS_PROFILE is interface-based.
+
+4. ARS_NEXTHOPS holds list of nexthops which are used for identifying ARS-enabled NHG. This table is applicable only when ars_nhg_path_selector_mode in ARS_PROFILE is nexthop.
```
From the example section:
"ARS_NEXTHOPS": {
    "default|1.1.1.10" : {
        "ars_obj_name": "ars_obj_name"
    },
    "default|2.2.2.20" : {
        "ars_obj_name": "ars_obj_name",
        "role": "alternative_path"
    },
    "default|3.3.3.30" : {
        "ars_obj_name": "ars_obj_name"
    }
}

I understand, this, as well as the ARS_INTERFACES are designed to apply ARS per nexthop/interface. Does SAI Support this? What does it mean to say that I have ARS operating in 'flowlet_mode' on interface A, and 'packet spray mode' on interface B?
Also, the implementation has ARS_NEXTHOP_GROUP table being used to match the route prefix. Is that still the primary mechanism to identify the NexthopGroup for ARS?
I am looking at a way where an ARS Object can be applied to a NexthopGroup that contains a given set of interfaces. Does your HLD address that?

baxia-lan commented on doc/ARS/Local_ARS_HLD.md:
```diff
@@ -0,0 +1,1151 @@
+                description "ARS-enabled interface name";
+            }
+
+            leaf scaling_factor {
```
the definition of "scaling_factor" differs from SAI_PORT_ATTR_ARS_PORT_LOAD_SCALING_FACTOR and brings ambiguity. 

`scaling_factor`: 10000
`SAI_PORT_ATTR_ARS_PORT_LOAD_SCALING_FACTOR`:40 # for 400G

Why not just set the value of SAI_PORT_ATTR_ARS_PORT_LOAD_SCALING_FACTOR to be value of scaling_factor?

`scaling_factor`: 40
`SAI_PORT_ATTR_ARS_PORT_LOAD_SCALING_FACTOR`:40 # for 400G
